src/main/java/hudson/remoting/Engine.java

```diff
@@ -361,8 +340,13 @@ private Socket connect(String host, String port) throws IOException, Interrupted
                 }
                 return s;
             } catch (IOException e) {
-                if(retry++>10)
-                    throw (IOException)new IOException("Failed to connect to "+host+':'+port).initCause(e);
+                if(retry++>10) {
+                    String suffix = "";
+                    if(isHttpProxy) {
+                        suffix = " through proxy " + targetAddress.toString();
+                    }
+                    throw (IOException) new IOException("Failed to connect to " + host + ':' + port + suffix).initCause(e);
```

Comment:
Sorry, initCause is returning a Throwable. Therefore we need to cast it to IOException.